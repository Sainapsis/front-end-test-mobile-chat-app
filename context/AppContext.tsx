import React, { createContext, useContext, ReactNode } from 'react';
import { useUser } from '@/hooks/useUser';
import { useChats } from '@/hooks/useChats';
import { DatabaseProvider } from '../database/DatabaseProvider';
import { useDatabase } from '@/hooks/useDatabase';
import { User } from '@/types/User';
import { Chat } from '@/types/Chat';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (chatId: string, text: string, senderId: string) => Promise<boolean>;
  loading: boolean;
  dbInitialized: boolean;
  clearChats: (userId: string) => Promise<void>;
  deleteChat?: (chatId: string, userId: string) => Promise<void>;
  deleteMessage?: (messageId: string, chatId: string) => Promise<boolean>;
  addReaction?: (messageId: string, emoji: string) => Promise<boolean>;
  removeReaction?: (reactionId: string, messageId: string) => Promise<boolean>;
  editMessage?: (messageId: string, newText: string) => Promise<boolean>;
};

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  const chatContext = useChats(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value: AppContextType = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
    deleteChat: chatContext.deleteChat || undefined,
    deleteMessage: chatContext.deleteMessage || undefined,
    addReaction: chatContext.addReaction || undefined,
    removeReaction: chatContext.removeReaction || undefined,
    editMessage: chatContext.editMessage || undefined,
  };

  if (!isInitialized) {
    return null;
  }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export const AppProvider = ({ children }: { children: React.ReactNode }) => (
  <DatabaseProvider>
    <AppContent>{children}</AppContent>
  </DatabaseProvider>
);

export function useAppContext() {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
}
