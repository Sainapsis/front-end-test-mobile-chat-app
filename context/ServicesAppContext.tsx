import React, { createContext, useContext, ReactNode } from 'react';
import { DatabaseProvider, useDatabaseStatus } from './DataBaseContext';
import { useChatsDb,  useUserDb } from '../hooks/db';
import { Chat, Message, User } from '@/utils/types';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (
    chatId: string, 
    text: string | undefined, 
    senderId: string,
    mediaUrl?: string,
    mediaType?: string,
    mediaThumbnail?: string,
    voiceUrl?: string,
    voiceDuration?: number,
    isVoiceMessage?: boolean,
    
  ) => Promise<Message | null>;
  addReaction: (messageId: string, userId: string, emoji: string) => Promise<boolean>;
  removeReaction: (reactionId: string) => Promise<boolean>;
  searchMessages: (query: string, chatId?: string) => Promise<void>;
  searchResults: Message[];
  deleteMessage: (messageId: string) => Promise<boolean>;
  editMessage: (messageId: string, newText: string) => Promise<boolean>;
  updateMessageStatus: (messageId: string, status: 'sent' | 'delivered' | 'read') => Promise<void>;
  markMessagesAsRead: (chatId: string) => Promise<void>;
  loading: boolean;
  dbInitialized: boolean;
};

export const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabaseStatus();
  const userContext = useUserDb();
  const chatContext = useChatsDb(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ children }: { children: ReactNode }) {
  return (
    <DatabaseProvider>
      <AppContent>{children}</AppContent>
    </DatabaseProvider>
  );
}

