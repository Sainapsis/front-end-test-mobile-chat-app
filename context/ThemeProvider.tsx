import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ThemeContext from './ThemeContext';
import { Theme } from '@/types/tColores';

/**
 * Provider component that handles theme state and persistence
 * @param children - Child components to be wrapped by the provider
 */
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as Theme);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = async () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    try {
      await AsyncStorage.setItem('theme', newTheme);
    } catch (error) {
      console.error('Error saving theme:', error);
    }
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};