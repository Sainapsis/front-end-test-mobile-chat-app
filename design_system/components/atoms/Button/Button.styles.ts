import { StyleSheet, ViewStyle, TextStyle } from 'react-native';
import { spacing, radius, typography, themes } from '@/design_system/ui/tokens';
import { Theme } from '@/types/tColores';

export const createButtonStyles = (theme: Theme) => StyleSheet.create({
  base: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: radius.md,
  } as ViewStyle,

  primary: {
    backgroundColor: themes[theme].button?.primary?.background,
    borderColor: themes[theme].button?.primary?.border,
  } as ViewStyle,

  secondary: {
    backgroundColor: themes[theme].button?.secondary.background,
    borderColor: themes[theme].button?.secondary.border,
  } as ViewStyle,

  outline: {
    backgroundColor: themes[theme].button?.outline.background,
    color: themes[theme].button?.outline.text,
    borderColor: themes[theme].button?.outline.border,
    borderWidth: 1,
  } as ViewStyle,

  ghost: {
    backgroundColor: themes[theme].button?.ghost.background,
    borderColor: themes[theme].button?.ghost.border,
  } as ViewStyle,

  small: {
    paddingVertical: spacing.xs,
    paddingHorizontal: spacing.md,
  } as ViewStyle,

  medium: {
    paddingVertical: spacing.sm,
    paddingHorizontal: spacing.lg,
  } as ViewStyle,

  large: {
    paddingVertical: spacing.md,
    paddingHorizontal: spacing.xl,
  } as ViewStyle,

  fullWidth: {
    width: '100%',
  } as ViewStyle,

  disabled: {
    backgroundColor: themes[theme].button?.disabled.background,
    borderColor: themes[theme].button?.disabled.border,
    opacity: 0.5,
  } as ViewStyle,

  text: {
    ...typography.variants.body1,
    textAlign: 'center',
  } as TextStyle,

  primaryText: {
    color: themes[theme].button?.primary?.text,
  } as TextStyle,

  secondaryText: {
    color: themes[theme].button?.secondary.text,
  } as TextStyle,

  outlineText: {
    color: themes[theme].button?.outline.text,
  } as TextStyle,

  ghostText: {
    color: themes[theme].button?.ghost.text,
  } as TextStyle,

  smallText: {
    fontSize: typography.sizes.sm,
    lineHeight: typography.lineHeights.normal,
  } as TextStyle,

  mediumText: {
    fontSize: typography.sizes.md,
    lineHeight: typography.lineHeights.normal,
  } as TextStyle,

  largeText: {
    fontSize: typography.sizes.lg,
  } as TextStyle,

  disabledText: {
    color: themes[theme].button?.disabled.text,
  } as TextStyle,

  contentContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    display: 'flex',
    gap: spacing.md,
  } as ViewStyle,

  icon: {
    marginHorizontal: spacing.xs,
  } as ViewStyle,
});