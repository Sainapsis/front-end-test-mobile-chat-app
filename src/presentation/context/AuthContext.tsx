import { LoginParams } from "@/src/data/interfaces/auth.interface";
import { createContext, ReactNode, useContext } from "react";
import { useAuth } from "../hooks/useAuth";
import { User } from "@/src/domain/entities/user";

interface AuthContextType {
  currentUser: User | null;
  login: ({ userId }: LoginParams) => Promise<boolean>;
  logout: () => void;
}

const _AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthContext({ children }: { children: ReactNode }) {
  const authContext = useAuth();

  const value = {
    ...authContext,
  };

  return (
    <_AuthContext.Provider value={value}>{children}</_AuthContext.Provider>
  );
}

export function useAuthContext() {
  const context = useContext(_AuthContext);

  if (context === undefined) {
    throw new Error("useAuthContext must be used within an AppProvider");
  }

  return context;
}
