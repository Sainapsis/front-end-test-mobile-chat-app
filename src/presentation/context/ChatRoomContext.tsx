import { DeleteMessageParams, EditMessageParams, UpdateStatusMessageParams } from '@/src/data/interfaces/chatRoom.interface';
import { Chat } from '@/src/domain/entities/chat';
import { Message, MessageStatus } from '@/src/domain/entities/message';
import { createContext, ReactNode, useContext } from "react";
import { useChatRoom } from '../hooks/useChatRoom';

interface ChatRoomContextType {
  updateStatus: (
    currentUserId: string,
    chat: Chat,
    statusMessage: MessageStatus
  ) => Promise<void>;
  sendMessage: (chatId: string, message: Message) => Promise<boolean>;
  editMessage: ({
    chatId,
    messageId,
    newText,
  }: EditMessageParams) => Promise<boolean>;
  deleteMessage: ({
    chatId,
    messageId,
  }: DeleteMessageParams) => Promise<boolean>;
  updateMessageStatus: (
    chatId: string,
    { messageId, status }: UpdateStatusMessageParams
  ) => Promise<void>;
  handleLoadMoreMessage: ({ chatId }: { chatId: string }) => Promise<void>;
}

const _ChatRoomContext = createContext<ChatRoomContextType | undefined>(undefined);

export function ChatRoomContext({ chatId, children }: { chatId: string; children: ReactNode }) {
  const chatRoomContext = useChatRoom({ chatId });

  const value = {
    ...chatRoomContext,
  };

  return (
    <_ChatRoomContext.Provider value={value}>{children}</_ChatRoomContext.Provider>
  );
}

export function useChatRoomContext() {
  const context = useContext(_ChatRoomContext);

  if (context === undefined) {
    throw new Error("useChatRoomContext must be used within an AppProvider");
  }

  return context;
}
