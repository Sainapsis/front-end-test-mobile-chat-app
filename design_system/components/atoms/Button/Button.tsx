import React from 'react';
import { ViewStyle, TextStyle, ActivityIndicator, View, StyleProp } from 'react-native';
import { ThemedText } from '@/design_system/components/atoms/ThemedText';
import { AnimatedPressable } from '@/design_system/components/atoms/Pressable';
import { colors } from '@/design_system/ui/tokens';
import { createButtonStyles } from "./Button.styles";
import { useTheme } from '@/context/ThemeContext';

type ButtonVariant = 'primary' | 'secondary' | 'outline' | 'ghost';
type ButtonSize = 'small' | 'medium' | 'large';

interface ButtonProps {
  /** Function to be called when the button is pressed */
  onPress?: () => void;
  /** Determines the style of the button */
  variant?: ButtonVariant;
  /** Determines the size of the button */
  size?: ButtonSize;
  /** If true, the button is disabled */
  disabled?: boolean;
  /** If true, an activity indicator is shown instead of the button content */
  loading?: boolean;
  /** If true, the button spans the full width of its container */
  fullWidth?: boolean;
  /** Icon to be displayed on the left side of the button content */
  leftIcon?: React.ReactNode;
  /** Icon to be displayed on the right side of the button content */
  rightIcon?: React.ReactNode;
  /** Content to be displayed inside the button */
  children: React.ReactNode;
  /** Custom styles to be applied to the button */
  style?: StyleProp<ViewStyle>;
}

/**
 * Button component used to trigger actions in the application.
 * Supports various styles, sizes, and states to fit different use cases.
 */
export function Button({
  onPress,
  variant = 'primary',
  size = 'medium',
  disabled,
  loading,
  fullWidth,
  leftIcon,
  rightIcon,
  children,
  style,
}: ButtonProps) {
  const { theme } = useTheme();
  const styles = createButtonStyles(theme);

  const buttonStyles: StyleProp<ViewStyle> = [
    styles.base,
    styles[variant],
    styles[size],
    fullWidth && styles.fullWidth,
    disabled && styles.disabled,
    style as ViewStyle,
  ].filter(Boolean);

  const textStyles: StyleProp<TextStyle> = [
    styles.text,
    styles[`${variant}Text`],
    styles[`${size}Text`],
    disabled && styles.disabledText,
  ].filter(Boolean);

  const renderContent = () => {
    if (typeof children === 'string') {
      return <ThemedText style={[textStyles]}>{children}</ThemedText>;
    }
    return children;
  };

  return (
    <AnimatedPressable
      onPress={onPress}
      disabled={disabled || loading}
      style={buttonStyles}
    >
      {loading ? (
        <ActivityIndicator
          color={variant === 'primary' ? colors.border.default : colors.primary.dark}
        />
      ) : (
        <View style={styles.contentContainer}>
          {Boolean(leftIcon) && <View style={styles.icon}>{leftIcon}</View>}
          {renderContent()}
          {Boolean(rightIcon) && <View style={styles.icon}>{rightIcon}</View>}
        </View>
      )}
    </AnimatedPressable>
  );
}
