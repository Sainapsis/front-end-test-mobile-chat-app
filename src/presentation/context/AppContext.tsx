import React, { createContext, useContext, ReactNode } from "react";
import { useDatabase } from "../hooks/useDatabase";
import { useUser } from "@/src/presentation/hooks/useUser";
import { DatabaseContext } from "@/src/presentation/context/DatabaseProvider";
import { UserContext } from "./UserContext";
import { AuthContext } from "./AuthContext";

interface AppContextType {
  dbInitialized: boolean;
}

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  
  const value = {
    ...userContext,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ chatId, children }: { chatId: string; children: ReactNode }) {
  return (
    <DatabaseContext>
      <AuthContext>
        <UserContext>
          <AppContent>{children}</AppContent>
        </UserContext>
      </AuthContext>
    </DatabaseContext>
  );
}

export function useAppContext() {
  const context = useContext(AppContext);

  if (context === undefined) {
    throw new Error("useAppContext must be used within an AppProvider");
  }

  return context;
}
