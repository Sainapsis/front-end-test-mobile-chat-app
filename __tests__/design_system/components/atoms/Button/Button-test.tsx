import * as React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { Button } from '@/design_system/components/atoms/Button/Button';
import { View, Text } from 'react-native';

// Mocks mÃ¡s simples que funcionan con Jest
jest.mock('@/design_system/components/atoms/ThemedText', () => {
    const { Text } = require('react-native'); // Importar dentro del mock
    return {
      ThemedText: ({ children, ...props }: { children: React.ReactNode; [key: string]: any }) => (
        <Text {...props}>{children}</Text>
      ),
    };
  });
  
  jest.mock('@/design_system/components/atoms/Pressable', () => {
    const { View } = require('react-native'); // Importar dentro del mock
    return {
      AnimatedPressable: ({ children, onPress, ...props }: { children: React.ReactNode; onPress?: () => void; [key: string]: any }) => (
        <View testID="animated-pressable" {...props}>
          {children}
        </View>
      ),
    };
  });
  

describe('Button Component', () => {
  const renderWithTheme = (component: string | number | boolean | Iterable<React.ReactNode> | React.JSX.Element | null | undefined) => {
    return render(
      <ThemeProvider>
        {component}
      </ThemeProvider>
    );
  };

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <Button>Test Button</Button>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders text content correctly', () => {
    const { getByText } = renderWithTheme(
      <Button>Test Button</Button>
    );
    expect(getByText('Test Button')).toBeTruthy();
  });

  it('handles onPress events', () => {
    const onPressMock = jest.fn();
    const { getByTestId } = renderWithTheme(
      <Button onPress={onPressMock}>Clickable Button</Button>
    );
    
    const buttonElement = getByTestId('animated-pressable');
    fireEvent.press(buttonElement);
    
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it('renders with different variants', () => {
    const variants = ['primary', 'secondary', 'outline', 'ghost'] as const;
    
    variants.forEach(variant => {
      const { getByText } = renderWithTheme(
        <Button variant={variant}>{`${variant} button`}</Button>
      );
      expect(getByText(`${variant} button`)).toBeTruthy();
    });
  });

  it('renders with different sizes', () => {
    const sizes = ['small', 'medium', 'large'] as const;
    
    sizes.forEach(size => {
      const { getByText } = renderWithTheme(
        <Button size={size}>{`${size} button`}</Button>
      );
      expect(getByText(`${size} button`)).toBeTruthy();
    });
  });

  // Resto de las pruebas sin cambios
});