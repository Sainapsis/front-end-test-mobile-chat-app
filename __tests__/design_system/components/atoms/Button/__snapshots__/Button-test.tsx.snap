// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly with snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#2CC5C0FF",
        "borderColor": "#0F1631FF",
      },
      {
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      },
    ]
  }
  testID="animated-pressable"
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "gap": 12,
      }
    }
  >
    <Text
      style={
        [
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 1.5,
              "textAlign": "center",
            },
            {
              "color": "#FFFFFF",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
          ],
        ]
      }
    >
      Test Button
    </Text>
  </View>
</View>
`;
